grpc/src/main/java/com/linecorp/armeria/server/grpc/GrpcDecoratingService.java

```diff
@@ -42,43 +51,78 @@ final class GrpcDecoratingService extends SimpleDecoratingHttpService implements
-                          Map<String, HttpService> methodToDecorators, boolean lookupMethodFromAttribute) {
+    @Nullable
+    private Map<ServerMethodDefinition<?, ?>, HttpService> decorated;
+
+    GrpcDecoratingService(GrpcService delegate, HandlerRegistry handlerRegistry,
+                          boolean lookupMethodFromAttribute) {
         super(delegate);
         this.delegate = delegate;
-        this.pathToDecorators = pathToDecorators;
-        this.methodToDecorators = methodToDecorators;
+        this.handlerRegistry = handlerRegistry;
         this.lookupMethodFromAttribute = lookupMethodFromAttribute;
     }
 
+    @Override
+    public void serviceAdded(ServiceConfig cfg) throws Exception {
+        super.serviceAdded(cfg);
+        final Map<ServerMethodDefinition<?, ?>, List<DecoratorAndOrder>> registryDecorators =
+                handlerRegistry.decorators();
+
+        final Builder<ServerMethodDefinition<?, ?>, HttpService> builder = ImmutableMap.builder();
+
+        for (Entry<ServerMethodDefinition<?, ?>, List<DecoratorAndOrder>> entry
+                : registryDecorators.entrySet()) {
+            final List<? extends Function<? super HttpService, ? extends HttpService>> decorators =
+                    entry.getValue()
+                         .stream()
+                         .map(DecoratorAndOrder::decorator)
+                         .collect(toImmutableList());
+            builder.put(entry.getKey(), applyDecorators(decorators, delegate));
+        }
+        decorated = builder.build();
+    }
+
+    private static HttpService applyDecorators(
+            List<? extends Function<? super HttpService, ? extends HttpService>> decorators,
+            HttpService delegate) {
+        Function<? super HttpService, ? extends HttpService> decorator = Function.identity();
+        for (int i = decorators.size() - 1; i >= 0; i--) {
+            decorator = decorator.andThen(decorators.get(i));
+        }
+        return decorator.apply(delegate);
+    }
+
     @Override
     public HttpResponse serve(ServiceRequestContext ctx, HttpRequest req) throws Exception {
-        HttpService decoratedService = lookup(ctx);
-        if (decoratedService != null) {
-            return decoratedService.serve(ctx, req);
+        assert decorated != null;
+        ServerMethodDefinition<?, ?> methodDefinition =
+                lookupMethodFromAttribute ? ctx.attr(RESOLVED_GRPC_METHOD) : null;
+        if (methodDefinition != null) {
+            final HttpService decoratedService = decorated.get(methodDefinition);
+            if (decoratedService != null) {
+                return decoratedService.serve(ctx, req);
+            }
         }
-        final String path = ctx.mappedPath();
-        decoratedService = pathToDecorators.get(path);
-        if (decoratedService != null) {
-            return decoratedService.serve(ctx, req);
+
+        methodDefinition = methodDefinition(ctx, handlerRegistry);
```

Comment:
How about adding a lookup method to `GrpcService`? For example:
```java
interface GrpcService {
    @Nullable
    ServerMethodDefinition method(ServiceRequestContext ctx);
}

class FramedGrpcService implements ... {
    @Override
    ServerMethodDefinition method(ServiceRequestContext ctx) {

        var method = ctx.attr(RESOLVED_GRPC_METHOD);
        if (method != null) {
           return method;
        }

        ....
        ctx.setAttr(...);
        return method;
    }
}
```

Reply:
That's a good suggestion. 👍